### PR TITLE
Modal 2.0

### PR DESCRIPTION
**Pull request by Marswang92**
# Change Log
1. Use `onOpen`, `onClose` props instead of  `beforeOpen`, `afterOpen`, `beforeClose`, `afterClose`.
2. `size` prop removed. Instead user can set modal size through `stylesheets` prop.
3. Remove `disableHeader` prop.
4. Unit test introduced.

## related issue
https://github.com/procore/particles/issues/317

### CODE REVIEW

File: modal/.babelrc [hunk 1]
```diff
@@ -0,0 +1,3 @@
+{
+  "presets": ["latest", "react", "stage-0"]
```
Comment (ben-burlingham) [inline modal/.babelrc]: Is this file still necessary?

File: modal/sandbox/sandbox.scss [hunk 2]
```diff
@@ -1,3 +1,9 @@
+// UX Approved Color Pallette
+$black1: #343739;
+$gray3: #c3c2c2;
+$gray4: #e7e7e7;
+$white1: #fff;
```
Comment (ben-burlingham) [inline modal/sandbox/sandbox.scss]: @nickwmurphy do these align with UX particles palette

File: modal/sandbox/sandbox.jsx [hunk 3]
```diff
@@ -21,9 +21,10 @@ class Component extends React.Component {
           closeModal={() => this.setState({ show: false })}
           size='large'
```
Comment (ben-burlingham) [inline modal/sandbox/sandbox.jsx]: `size` was removed

File: modal/sandbox/sandbox.jsx [hunk 4]
```diff
@@ -21,9 +21,10 @@ class Component extends React.Component {
...
           slotHeader={'Header'}
-          slotFooter={(<div className={sandboxStyles.footer}>Footer</div>)}
-          onShow={show => this.setState({ show })}
-          onHide={show => this.setState({ show })}
+          slotFooter={(<div>Footer</div>)}
+          stylesheets={[sandboxStyles]}
+          onOpen={() => { console.log('open'); }}
+          onClose={() => { console.log('close'); }}
```
Comment (ben-burlingham) [inline modal/sandbox/sandbox.jsx]: nb The other sandboxes do these declarations differently, I think for consistency we should follow them but no big deal.  Maybe we should talk about what standards we want on the sandboxes.

File: modal/src/Modal.jsx [hunk 5]
```diff
@@ -9,27 +10,19 @@ const isClosing = (s1, s2) => s1.show && !s2.show;
...
   componentDidUpdate(prevProps) {
-    if (isOpening(prevProps, this.props) && this.props.afterOpen) {
-      this.props.afterOpen();
+    if (isOpening(prevProps, this.props) && this.props.onOpen) {
+      this.props.onOpen();
     }
 
-    if (isClosing(prevProps, this.props) && this.props.afterClose) {
-      this.props.afterClose();
+    if (isClosing(prevProps, this.props) && this.props.onClose) {
```
Comment (ben-burlingham) [inline modal/src/Modal.jsx]: Shouldn't need `&& this.props.onClose` if default props have declared it to `() => {}`

File: modal/src/Modal.jsx [hunk 6]
```diff
@@ -43,67 +36,55 @@ class Modal extends Component {
...
-        [styles['modal-small']]: size === 'small',
-        [styles['modal-medium']]: size === 'medium',
-        [styles['modal-large']]: size === 'large'
-      }
-    );
     const closeClassName = cx(
-      styles.close, {
-        [styles['close-black']]: disableHeader
+      this.styles.close, {
+        [this.styles['close-black']]: !slotHeader
```
Comment (ben-burlingham) [inline modal/src/Modal.jsx]: nb This feels funny.  Seems like there should be one close style.  I see what you're going for here though...maybe we can iterate in the future.

File: modal/src/Modal.jsx [hunk 7]
```diff
@@ -43,67 +36,55 @@ class Modal extends Component {
...
-          )}
-          <div className={styles.content}>
-            { children }
+      <div className={this.styles.container}>
+        <div className={this.styles['modal-window']}>
+          <div className={this.styles['modal-header-container']}>
+            <div className={closeClassName}>
+              <i
+                className='fa fa-close'
+                onClick={this.props.closeModal}
```
Comment (ben-burlingham) [inline modal/src/Modal.jsx]: Shouldn't this call `onClose` from the props?

File: modal/src/Modal.jsx [hunk 8]
```diff
@@ -120,35 +101,25 @@ Modal.propTypes = {
...
-  afterClose: PropTypes.func,
-  onCloseClicked: PropTypes.func,
+  onOpen: PropTypes.func,
+  onClose: PropTypes.func,
   closeModal: PropTypes.func
 };
 
 Modal.defaultProps = {
   stylesheets: [],
   children: <div />,
```
Comment (ben-burlingham) [inline modal/src/Modal.jsx]: IMO `slotBody` would be more appropriate since follows other props

File: modal/sandbox/sandbox.jsx (hunk 4 above)
Comment (ben-burlingham) [reply modal/sandbox/sandbox.jsx]: ^ @mboperator

File: modal/.babelrc (hunk 1 above)
Comment (mcclayton) [reply modal/.babelrc]: With Jest, yes.
https://github.com/facebook/jest#using-babel
There may be a way around it, but a `.babelrc` file is the recommended approach.

File: modal/test/Modal.test.jsx [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+  const mergedProps = Object.assign({}, props, newProps);
+  return mount(
+    <Modal {...mergedProps} />
+  );
+};
+
+let showWrapper;
+let hideWrapper;
+
+beforeEach(() => {
```
Comment (mcclayton) [inline modal/test/Modal.test.jsx]: `nb`
I don't believe this needs to be in a `beforeEach`.
You can simply create two consts:
```
const showWrapper = renderModal({ show: true });
const hideWrapper = renderModal({ show: false });
```

File: modal/src/Modal.jsx (hunk 7 above)
Comment (Marswang92) [reply modal/src/Modal.jsx]: yes mb

Comment (Marswang92) [reply modal/src/Modal.jsx]: Cuz previously there's a callback specifically for user click on the close button

File: modal/src/Modal.jsx (hunk 6 above)
Comment (Marswang92) [reply modal/src/Modal.jsx]: I think proper way to do it is just let user specify the style of the close button. Anyway we can figure out according to the people's feedback on using this.
